lib/ddtrace/configuration/components.rb

```diff
@@ -54,28 +54,116 @@ def build_tracer(settings, agent_settings)
           tracer = settings.tracer.instance
           return tracer unless tracer.nil?
 
-          tracer = Tracer.new(
+          # Apply test mode settings if test mode is activated
+          if settings.test_mode.enabled
+            context_flush = build_test_mode_context_flush(settings)
+            sampler = build_test_mode_sampler
+            writer = build_test_mode_writer(settings, agent_settings)
+          else
+            context_flush = build_context_flush(settings)
+            sampler = build_sampler(settings)
+            writer = build_writer(settings, agent_settings)
+          end
+
+          subscribe_to_writer_events!(writer, sampler)
```

Comment:
To be honest, I had left here a comment suggesting that some of this behavior be moved to the `Sampler` or the `Writer`, and then reading below I realized that this PR actually is going in the other direction :)

I'm... not entirely convinced lol. The interactions between tracer, sampler and writer are already complex enough, that I think this abstraction makes the whole thing even more hard to follow.

Would you consider breaking this change off from this PR and into its own? That would allow us to focus on that interaction separately from rewiring how all three classes get initialized.

Reply:
> Would you consider breaking this change off from this PR and into its own? That would allow us to focus on that interaction separately from rewiring how all three classes get initialized.

Break off the initialization changes of the Writer and Sampler from removing `Tracer#configure`?

I could do the initialization changes of the Writer and Sampler before removing `Tracer#configure`, but not in parallel. They would also overlap a bit: I'd have to write a stop-gap solution for the `Writer#update_priority_sampler` change, reconciling the old way of providing the `priority_sampler` to the Writer, while maintaining it at `Tracer#configure` and write tests to account for both cases and their conflicts, and then remove that stop-gap solution in the `Tracer#configure` PR.

If I break it down like suggested, it would make reviewing easier (unless the stop gap solutions turn out uglier and make reviewing the intermediate PRs worse), but it won't affect the end result. Given I can't even think of an easy way to gracefully break down the `Writer.@priority_sampler` dependency without making pretty much most of the changes in this PR, I'm inclined to say that I don't think breaking it down like suggested is a good idea.

Reply:
> Break off the initialization changes of the Writer and Sampler from removing Tracer#configure?
> 
> I could do the initialization changes of the Writer and Sampler before removing Tracer#configure, but not in parallel. They would also overlap a bit: I'd have to write a stop-gap solution for the Writer#update_priority_sampler change, reconciling the old way of providing the priority_sampler to the Writer, while maintaining it at Tracer#configure and write tests to account for both cases and their conflicts, and then remove that stop-gap solution in the Tracer#configure PR.

Hmmm this may bit more complex than what I was thinking, but that wasn't quite what I was suggesting.

The thing I'm not quite convinced is the right direction and would like to look at separately is the introduction of events into `Writer` and `SyncWriter` as an indirect way of connecting them to the `Sampler`.

That's because I don't see it as a very generic abstraction that we'll end up using for lots of things, but as a generic abstraction which we'll use one for this one thing -- connecting the writers to the sampler -- and thus I fear it'll be one more layer of indirection added to something that's already very complex and hard to follow.